core/src/main/java/smile/projection/ICA.java

```diff
@@ -0,0 +1,365 @@
+                    wp[j] /= norm;
+                }
+
+                // step 3 check if convergence
+                diff = Math.abs(Math.abs(Math.dot(wp, lastwp)) - 1);
+
+                lastwp = wp;
+
+                iterCnt++;
+                if (iterCnt > this.maxIter) {
+                    throw new IllegalArgumentException(
+                            String.format("fail to find the No.%d Independent Component within given iterations: %d",
+                                    (i + 1), maxIter));
+                }
+            }
+            wps[i] = lastwp;
+            for (int j = 0; j < n; j++) {
+                projection.set(i, j, lastwp[j]);
+            }
+        }
+
+        pmu = Math.colMeans(projection.transpose().array());
+    }
+
+    @Override
+    public double[] project(double[] x) {
+        return project(new double[][] { x })[0];
+    }
+
+    @Override
+    public double[][] project(double[][] x) {
+        if (x[0].length != n) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid input vector size: %d, expected: %d", x[0].length, n));
+        }
+
+        double[][] y = new double[x.length][p];
+        DenseMatrix whitened = whiteData(x);
+        double[][] data = whitened.array();
+        for (int i = 0; i < data.length; i++) {
+            projection.ax(data[i], y[i]);
+            Math.minus(y[i], pmu);
+        }
+        return y;
+    }
+
+    /**
+     * 
+     * @return the projection matrix
+     */
+    public DenseMatrix getProjection() {
+        return projection;
+    }
+
+    /**
+     * 
+     * @return the functor {@link NegEntropyFunc} to get first or second order
+     *         derivative
+     */
+    public NegEntropyFunc getFuncMode() {
+        return funcMode;
+    }
+
+    /**
+     * Before applying ICA, we need to do whitening against the raw dataset.
+     * 
+     * @param data
+     * @return whitened dataset
+     */
+    private DenseMatrix whiteData(double[][] data) {
+        int m = data.length;
+        // center data
+        double[] mu = Math.colMeans(data);
+        DenseMatrix centered = Matrix.newInstance(data);
+        for (int j = 0; j < n; j++) {
+            for (int i = 0; i < m; i++) {
+                centered.sub(i, j, mu[j]);
+            }
+        }
+
+        // covariance matrix on centered data.
+        DenseMatrix T = Matrix.zeros(n, n);
+        for (int i = 0; i < m; i++) {
+            for (int j = 0; j < n; j++) {
+                for (int l = 0; l <= j; l++) {
+                    T.add(j, l, centered.get(i, j) * centered.get(i, l));
+                }
+            }
+        }
+        for (int j = 0; j < n; j++) {
+            for (int l = 0; l <= j; l++) {
+                T.div(j, l, m);
+                T.set(l, j, T.get(j, l));
+            }
+        }
+
+        T.setSymmetric(true);
+        EVD eigen = T.eigen();
+        DenseMatrix D = eigen.getEigenVectors();
```

Comment:
Why `D` not `E` as in the wiki link you post? Keeping variable names consistent with your reference will make it easy to check the formula.